apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/strategy/ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSet.java

```diff
@@ -132,29 +129,32 @@ protected ApplyNovaTemplateOptionsCreateNodesWithGroupEncodedIntoNameThenAddToSe
       ImmutableList.Builder<String> tagsBuilder = ImmutableList.builder();
 
       if (!templateOptions.getGroups().isEmpty()) {
-         Set<String> securityGroupNames = novaApi.getSecurityGroupApi(region).get().list()
-                 .transform(new Function<SecurityGroup, String>() {
-                    @Override
-                    public String apply(SecurityGroup input) {
-                       return input.getName();
-                    }
-                 })
-                 .toSet();
+         Iterable<String> securityGroupNames = Iterables.transform(securityGroupExtension.listSecurityGroups(), new Function<org.jclouds.compute.domain.SecurityGroup, String>() {
+            @Override
+            public String apply(@Nullable org.jclouds.compute.domain.SecurityGroup input) {
+               return input.getName();
+            }
+         });
          for (String securityGroupName : templateOptions.getGroups()) {
-            checkState(securityGroupNames.contains(securityGroupName), "Cannot find security group with name " + securityGroupName + ". \nSecurity groups available are: \n" + Iterables.toString(securityGroupNames)); // {
+            checkState(Iterables.contains(securityGroupNames, securityGroupName), "Cannot find security group with name " + securityGroupName + ". \nSecurity groups available are: \n" + Iterables.toString(securityGroupNames)); // {
          }
-      }
-      else if (!inboundPorts.isEmpty()) {
-         SecurityGroupInRegion securityGroupInRegion;
+
+      } else if (!inboundPorts.isEmpty()) {
+         SecurityGroup securityGroup = null;
          String securityGroupName = namingConvention.create().sharedNameForGroup(group);
-         try {
-            securityGroupInRegion = securityGroupCache.get(new RegionSecurityGroupNameAndPorts(region, securityGroupName, inboundPorts));
-         } catch (ExecutionException e) {
-            throw Throwables.propagate(e.getCause());
+
+         // populate the security group cache with existing security groups
+         for (SecurityGroup existingSecurityGroup : securityGroupExtension.listSecurityGroupsInLocation(template.getLocation())) {
+            securityGroupCache.put(new RegionSecurityGroupNameAndPorts(region, existingSecurityGroup.getName(), inboundPorts), existingSecurityGroup);
          }
-         templateOptions.securityGroups(securityGroupName);
-         tagsBuilder.add(String.format("%s-%s", JCLOUDS_SG_PREFIX, securityGroupInRegion.getSecurityGroup().getId()));
+
+         if (securityGroup == null) {
```

Comment:
At this point this is always null. Remove the dedundant `if` and declare the variable here.